File: eggroll/storage/storage-service-java/src/main/java/com/webank/ai/eggroll/framework/storage/service/server/LMDBServicer.java
/*
 * Copyright 2019 The FATE Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.webank.ai.eggroll.framework.storage.service.server;


import com.google.protobuf.ByteString;
import com.webank.ai.eggroll.api.storage.KVServiceGrpc;
import com.webank.ai.eggroll.api.storage.Kv;
import com.webank.ai.eggroll.core.api.grpc.server.GrpcServerWrapper;
import com.webank.ai.eggroll.core.constant.MetaConstants;
import com.webank.ai.eggroll.core.io.KeyValue;
import com.webank.ai.eggroll.core.io.KeyValueIterator;
import com.webank.ai.eggroll.core.io.StoreInfo;
import com.webank.ai.eggroll.core.io.StoreManager;
import com.webank.ai.eggroll.core.model.Bytes;
import com.webank.ai.eggroll.core.serdes.impl.POJOUtils;
import com.webank.ai.eggroll.framework.storage.service.model.LMDBStore;
import io.grpc.*;
import io.grpc.stub.StreamObserver;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;


public class LMDBServicer extends KVServiceGrpc.KVServiceImplBase {
    private static final long PAYLOAD_THRESHOLD = 2L * 1024 * 1024;
    private static Logger LOGGER = LogManager.getLogger(LMDBServicer.class);
    private final StoreManager<Bytes, byte[]> storeMgr;
    private GrpcServerWrapper grpcServerWrapper;

    public LMDBServicer(StoreManager<Bytes, byte[]> storeMgr) {
        this.storeMgr = storeMgr;
        this.grpcServerWrapper = new GrpcServerWrapper();
    }

    private LMDBStore getStore() {
        StoreInfo info = StoreInfo.fromGrpcContext();
        return (LMDBStore) storeMgr.createIfMissing(info);
    }

    @Override
    public void put(Kv.Operand request, StreamObserver<Kv.Empty> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            LMDBStore store = getStore();
            LOGGER.info("{} receive put request. keyLength: {}, valueLength: {}",
                    store, request.getKey().size(), request.getValue().size());
            store.put(Bytes.wrap(request.getKey()), request.getValue().toByteArray());
            responseObserver.onNext(Kv.Empty.newBuilder().build());
            responseObserver.onCompleted();
        });
    }

    @Override
    public void putIfAbsent(Kv.Operand request, StreamObserver<Kv.Operand> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            LMDBStore store = getStore();
            byte[] oldValue = store.putIfAbsent(Bytes.wrap(request.getKey()), request.getValue().toByteArray());
            Kv.Operand.Builder builder = Kv.Operand.newBuilder().setKey(request.getKey());
            if (oldValue != null) {
                builder.setValue(ByteString.copyFrom(oldValue));
            }
            responseObserver.onNext(builder.buildPartial());
            responseObserver.onCompleted();
        });
    }

    @Override
    public StreamObserver<Kv.Operand> putAll(StreamObserver<Kv.Empty> responseObserver) {
        LMDBStore store = getStore();
        LOGGER.info("putAll request received. store: {}", store);
        StreamObserver<KeyValue<Bytes, byte[]>> putObs = store.putAll();

        return new StreamObserver<Kv.Operand>() {
            long count = 0;

            @Override
            public void onNext(Kv.Operand operand) {
                putObs.onNext(POJOUtils.buildKeyValue(operand));
                count++;
            }

            @Override
            public void onError(Throwable throwable) {
                LOGGER.error(throwable.getMessage(), throwable);
                putObs.onError(throwable);
                LOGGER.info(String.format("%s put all %d entries before error", store.toString(), count));
            }

            @Override
            public void onCompleted() {
                putObs.onCompleted();
                responseObserver.onNext(Kv.Empty.newBuilder().build());
                responseObserver.onCompleted();
                LOGGER.info(String.format("%s put all %d entries", store.toString(), count));
            }
        };
    }

    @Override
    public void delOne(Kv.Operand request, StreamObserver<Kv.Operand> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            LMDBStore store = getStore();

            byte[] valueBuffer = store.delete(Bytes.wrap(request.getKey()));
            Kv.Operand.Builder builder = Kv.Operand.newBuilder().setKey(request.getKey());
            if (valueBuffer != null) {
                builder.setValue(ByteString.copyFrom(valueBuffer));
            }
            responseObserver.onNext(builder.buildPartial());
            responseObserver.onCompleted();
        });
    }

    @Override
    public void get(Kv.Operand request, StreamObserver<Kv.Operand> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            LMDBStore store = getStore();
            byte[] valueBuffer = store.get(Bytes.wrap(request.getKey()));
            Kv.Operand.Builder builder = Kv.Operand.newBuilder().setKey(request.getKey());
            if (valueBuffer != null)
                builder.setValue(ByteString.copyFrom(valueBuffer));
            responseObserver.onNext(builder.buildPartial());
            responseObserver.onCompleted();
        });
    }

    @Override
    public void iterate(Kv.Range request, StreamObserver<Kv.Operand> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            long bytesCount = 0;
            LMDBStore store = getStore();
            LOGGER.info("{} receive iterate request. start: {}, end: {}, minChunkSize: {}",
                    store, request.getStart().toStringUtf8(), request.getEnd().toStringUtf8(), request.getMinChunkSize());

            byte[] fromBytes = request.getStart().toByteArray();
            byte[] toBytes = request.getEnd().toByteArray();
            Bytes fromBuffer = null;
            Bytes toBuffer = null;
            long count = 0;
            if (fromBytes.length > 0) {
                fromBuffer = Bytes.wrap(fromBytes);
            }
            if (toBytes.length > 0) {
                toBuffer = Bytes.wrap(toBytes);
            }
            long threshold = request.getMinChunkSize() > 0 ? request.getMinChunkSize() : PAYLOAD_THRESHOLD;
            try (KeyValueIterator<Bytes, byte[]> keyValueIterator = store.range(fromBuffer, toBuffer)) {
                while (keyValueIterator.hasNext()) {
                    KeyValue<Bytes, byte[]> keyValue = keyValueIterator.next();
                    Kv.Operand operand = Kv.Operand.newBuilder().setKey(ByteString.copyFrom(keyValue.key.get()))
                            .setValue(ByteString.copyFrom(keyValue.value)).build();
                    responseObserver.onNext(operand);
                    ++count;
                    bytesCount += operand.getKey().size();
                    bytesCount += operand.getValue().size();
                    if (bytesCount >= threshold) {
                        break;
                    }
                }
            } finally {
                responseObserver.onCompleted();
                LOGGER.info("[STORAGE][ITERATE][RESULT]: {} store count: {}, onNext count: {}, start: '{}', end: '{}'",
                        store, store.count(), count, request.getStart().toStringUtf8(), request.getEnd().toStringUtf8());
            }
        });
    }

    @Override
    public void destroy(Kv.Empty request, StreamObserver<Kv.Empty> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            StoreInfo info = StoreInfo.fromGrpcContext();
            storeMgr.destroy(info);
            responseObserver.onNext(Kv.Empty.newBuilder().build());
            responseObserver.onCompleted();
        });
    }

    @Override
    public void count(Kv.Empty request, StreamObserver<Kv.Count> responseObserver) {
        grpcServerWrapper.wrapGrpcServerRunnable(responseObserver, () -> {
            LMDBStore lmdbStore = getStore();
            responseObserver.onNext(Kv.Count.newBuilder().setValue(lmdbStore.count()).build());
            responseObserver.onCompleted();
        });
    }


    public static class KvStoreInterceptor implements ServerInterceptor {
        @Override
        public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> serverCall, Metadata metadata, ServerCallHandler<ReqT, RespT> serverCallHandler) {
            Context ctx = MetaConstants.updateContext(metadata, Context.current());
            return Contexts.interceptCall(ctx, serverCall, metadata, serverCallHandler);
        }
    }


}

File: arch/eggroll/storage-service-cxx/src/LMDBServicer.cc
/*
 * Copyright 2019 The FATE Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "LMDBServicer.h"


using std::cout;
using std::endl;
using std::string;

LMDBServicer::LMDBServicer() : LMDBServicer::LMDBServicer("/tmp") {}

LMDBServicer::LMDBServicer(std::string dataDir) {
    this->dataDir = dataDir;
}

LMDBServicer::~LMDBServicer() {

}

Status defaultStatus = Status::OK;

// put an entry to table
Status LMDBServicer::put(ServerContext *context, const Operand *request, Empty *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[PUT] request" << endl;
        cout << "[PUT] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        lmdbStore.put(request);

        LOG(INFO) << "[PUT] finished" << endl;
        cout << "[PUT] finished" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::put");

    return Status::OK;
}

// put an entry to table if absent
Status LMDBServicer::putIfAbsent(ServerContext *context, const Operand *request, Operand *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[PUTIFABSENT] request" << endl;
        cout << "[PUTIFABSENT] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        string_view oldValue = lmdbStore.putIfAbsent(request);

        response->set_key(request->key());
        response->set_value(oldValue.data(), oldValue.size());

        LOG(INFO) << "[PUTIFABSENT] finished" << endl;
        cout << "[PUTIFABSENT] finished" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::putIfAbsent");

    return Status::OK;
}

// put entries to table (entries will be streaming in)
Status LMDBServicer::putAll(ServerContext *context, ServerReader<Operand> *reader, Empty *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[PUTALL] request" << endl;
        cout << "[PUTALL] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        long i = lmdbStore.putAll(reader);

        LOG(INFO) << "[PUTALL] Finished. total: " << i << endl;
        cout << "[PUTALL] Finished. total: " << i << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::putAll");

    return Status::OK;
}

// delete an entry from table
Status LMDBServicer::delOne(ServerContext *context, const Operand *request, Operand *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[DELONE] request" << endl;
        cout << "[DELONE] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        string_view oldValue = lmdbStore.delOne(request);

        response->set_key(request->key());
        response->set_value(oldValue.data(), oldValue.size());

        LOG(INFO) << "[DELONE] finished" << endl;
        cout << "[DELONE] finished" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::delOne");

    return Status::OK;
}

// get an entry from table
Status LMDBServicer::get(ServerContext *context, const Operand *request, Operand *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[GET] request" << endl;
        cout << "[GET] request" << endl;

        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        string_view value = lmdbStore.get(request);

        response->set_key(request->key());
        response->set_value(value.data(), value.size());

        LOG(INFO) << "[GET] finished" << endl;
        cout << "[GET] finished" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::get");

    return Status::OK;
}

// iterate through a table. Response entries are ordered
Status LMDBServicer::iterate(ServerContext *context, const Range *request, ServerWriter<Operand> *writer) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[ITERATE] request" << endl;
        cout << "[ITERATE] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        lmdbStore.iterate(request, writer);

        LOG(INFO) << "[ITERATE] finished" << endl;
        cout << "[ITERATE] finished" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::iterate");

    return Status::OK;
}

// destroy a table
Status LMDBServicer::destroy(ServerContext *context, const Empty *request, Empty *response) {
    std::exception_ptr eptr;

    bool result = false;
    try {
        LOG(INFO) << "[DESTROY] request" << endl;
        cout << "[DESTROY] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        result = lmdbStore.destroy();

        LOG(INFO) << "[DESTROY] finished. result:" << result << endl;
        cout << "[DESTROY] finished. result: " << result << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::destroy");

    return Status::OK;
}

Status LMDBServicer::destroyAll(ServerContext *context, const Empty *request, Empty *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[DESTROYALL] request" << endl;
        cout << "[DESTROYALL] request" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::destroyAll");

    return Status::OK;
}

// count record amount of a table
Status LMDBServicer::count(ServerContext *context, const Empty *request, Count *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[COUNT] request" << endl;
        cout << "[COUNT] request" << endl;
        LMDBStore lmdbStore;
        getStore(context, lmdbStore);

        response->set_value(lmdbStore.count());

        LOG(INFO) << "[COUNT] finished" << endl;
        cout << "[COUNT] finished" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::count");

    return Status::OK;
}

Status LMDBServicer::createIfAbsent(ServerContext *context, const CreateTableInfo *request, CreateTableInfo *response) {
    std::exception_ptr eptr;
    try {
        LOG(INFO) << "[CREATEIFABSENT] request" << endl;
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::createIfAbsent");

    return Status::OK;
}

void LMDBServicer::sayHello() {
    LOG(INFO) << "saying hello" << endl;
}

void LMDBServicer::getStore(ServerContext* context, LMDBStore &lmdbStore) {
    std::exception_ptr eptr;
    try {
        std::multimap <grpc::string_ref, grpc::string_ref> metadata = context->client_metadata();

        StoreInfo storeInfo(context);

        bool result = lmdbStore.init(dataDir, storeInfo);

        // todo: add exception if result is false
        if (!result) {
            std::string errorMsg = "Unable to init LMDBStore. please check error log.";
            throw std::runtime_error(errorMsg);
        }
    } catch (...) {
        eptr = std::current_exception();
    }
    handle_eptr(eptr, __FILE__, __LINE__, "LMDBServicer::getStore");
}